Clarify GitHub/Google auth for dev env.

vaidap opened the pull request:
To-do:
- [x] fix test for google auth (second commit)
- [x] update comments in `prod_settings_template.py` to be clearer & in line with the latest github/google interfaces.
- [x] Port into [readthedocs](http://zulip.readthedocs.io/en/latest/settings.html)?

smarx commented:
*Automated message from Dropbox CLA bot*

@vaidap, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on tools/run-dev.py:
```diff
@@ -118,6 +118,21 @@
+            "You are using the Google auth backend. Please make sure of the following:",
+            "- You have updated GOOGLE_OAUTH2_CLIENT_ID and "
+            "  GOOGLE_OAUTH2_CLIENT_SECRET settings.",
+            "- You have setup an Oauth2 client ID that allows redirects' ",
```
do we want to provide the URL here?

timabbott commented:
This is looking good!  I like the approach of extending error messages; rather than do it in `run-dev.py`, I might have think about having it be what happens when you click the button, under the theory that some folks may want to just be doing design work on the login pages with those buttons present, but not actually want to have to get keys from GitHub/Google.

Re: auth setup, I think it'd be reasonable to just write verbose docs for setting up authetication for testing in development somewhere in RTD and then link to that from `/devtools`.

vaidap commented:
* Added more verbose section to RTD and linked from devtools to it
* Fixed the Google test & linked to the docs from it 
* Reworded `prod_settings_template.py`'s comments as they are a bit out of date with the newest GitHub/Google interfaces (would appreciate a check on whether this is completely correct)
* When it comes to frontend work, I've just realised one can just fake the buttons by copy-pasting the html code from inbetween the eg. `{% if google_auth_enabled %}` & `{% endif %}` blocks into the inspector.

timabbott commented:
Thanks @vaidap, this looks great!  Merged.

I think a good follow-up for this might be to move the GitHub/Google setup errors from `run-dev.py` to instead be errors one gets when clicking the buttons; I think that's a better flow, because you may not think to restart the Zulip `run-dev.py` after changing settings and thus not see the error message at present.  Does that concept make sense?

vaidap commented:
that makes sense in how the flow would work, yes! 
Would the idea be to do a check before the form submission somehow, and give developers a warning? Or as soon as the buttons load?

Interestingly enough, the google error that i coded here only triggers if `GOOGLE_OAUTH2_CLIENT_ID` has been set but not the secret. If nothing has been set, it triggers the error
```
Traceback (most recent call last):
  File "./tools/run-dev.py", line 122, in <module>
    if not (settings.GOOGLE_OAUTH2_CLIENT_ID and
  File "/srv/zulip-venv/local/lib/python2.7/site-packages/django/conf/__init__.py", line 57, in __getattr__
    val = getattr(self._wrapped, name)
AttributeError: 'Settings' object has no attribute 'GOOGLE_OAUTH2_CLIENT_ID'
```

timabbott commented:
I think we need to add `'GOOGLE_OAUTH2_CLIENT_ID': None` in `DEFAULT_SETTINGS` to fix that.

See zulip.readthedocs.io/en/latest/settings.html for relevant docs.

I would do the check when the user clicks the button, aka in the whatever Django view code handles the button click.

vaidap commented:
Thanks for that, it worked! Submitted two PRs that cover these points.
